src/Adapter/InMemoryAdapter.php

```diff
@@ -135,18 +136,22 @@ public function loadReverse(
 
     public function loadEvents(
         StreamName $streamName,
-        array $metadata = [],
         int $fromNumber = 0,
-        int $count = null
+        int $count = null,
+        MetadataMatcher $metadataMatcher = null
     ): Iterator {
         if (! isset($this->streams[$streamName->toString()])) {
             return new ArrayIterator();
         }
 
+        if (null === $metadataMatcher) {
+            $metadataMatcher = new MetadataMatcher();
+        }
+
         $streamEvents = [];
 
         foreach ($this->streams[$streamName->toString()]['events'] as $index => $streamEvent) {
```

Comment:
Instead of using a foreach here we should use a filter method. The whole code would be more readable, since this spaghetti condition is just not readable. Also a method 'getEventsFromStream' in a StreamCollection would be nice.